src/methods/join.js

```diff
@@ -936,31 +937,41 @@ export class JoinImplementation {
         invariant(dcValue.length > 0);
         let elem0 = dcValue[0];
         if (elem0 instanceof Value) {
-          dc.value = dcValue.map(e => getAbstractValue((e: any), realm.intrinsics.empty));
+          dc.value = dcValue.map(e => {
+            return useD1
+              ? getAbstractValue((e: any), realm.intrinsics.empty)
+              : getAbstractValue(realm.intrinsics.empty, (e: any));
+          });
         } else {
           dc.value = dcValue.map(e => {
             let { $Key: key1, $Value: val1 } = (e: any);
-            let key3 = getAbstractValue(key1, realm.intrinsics.empty);
-            let val3 = getAbstractValue(val1, realm.intrinsics.empty);
+            let key3 = useD1
+              ? getAbstractValue(key1, realm.intrinsics.empty)
+              : getAbstractValue(realm.intrinsics.empty, key1);
+            let val3 = useD1
+              ? getAbstractValue(val1, realm.intrinsics.empty)
+              : getAbstractValue(realm.intrinsics.empty, val1);
             return { $Key: key3, $Value: val3 };
           });
         }
       } else {
         invariant(dcValue === undefined || dcValue instanceof Value);
-        dc.value = getAbstractValue(dcValue, realm.intrinsics.empty);
+        dc.value = useD1
+          ? getAbstractValue(dcValue, realm.intrinsics.empty)
+          : getAbstractValue(realm.intrinsics.empty, dcValue);
```

Comment:
That's the branch I was fixing but I assume others had the same problem. I don't know how to test other branches.